txpool/txpool.go

```diff
@@ -471,60 +468,164 @@ func (t *TxPool) ResetWithHeader(h *types.Header) {
+	// from any queues in the TxPool
+	for _, txn := range txnsInBlocks {
+		// Initially drop any txns that are leftover in the TxPool queues,
+		// but are submitted to the chain state
+		dropTxnCallback(txn)
+
+		// Attempt to update the next account nonce in the TxPool
+		// with the one in state if it's greater
+		wrapper := t.lockAccountQueue(txn.From, true)
+
+		// Grab the latest nonce from state
+		stateRoot := t.store.Header().StateRoot
+		stateNonce := t.store.GetNonce(stateRoot, txn.From)
+
+		// Check if the state nonce is greater than the nonce present in the txpool
+		// If so -> realign it to the state nonce
+		if stateNonce > wrapper.accountQueue.nextNonce {
+			t.logger.Debug(
+				fmt.Sprintf(
+					"World state nonce [%d] for account [%s] is > the TxPool nonce [%d]",
+					stateNonce,
+					txn.From.String(),
+					wrapper.accountQueue.nextNonce,
+				),
+			)
+			wrapper.accountQueue.nextNonce = stateNonce
+		}
+
+		// Reassign for easier handling
+		accountNextNonce := wrapper.accountQueue.nextNonce
+
+		// Since there have been state changes, the TxPool can still have hanging txns.
+		// Prune out all the now possibly low-nonce transactions in the account queue
+		for {
+			// Check if the nonce is lower than what the TxPool is expecting
+			lowestNonceTx := wrapper.accountQueue.txs.Peek()
+
+			if lowestNonceTx == nil || // There is nothing in the account specific queue
+				lowestNonceTx.Nonce >= accountNextNonce { // The lowest nonce tx is valid
+				wrapper.accountQueue.logger.Debug(
+					fmt.Sprintf(
+						"All account-based txns for address [%s] are valid",
+						txn.From.String(),
+					),
+				)
+				// All good
+				break
+			}
+
+			// Remove the transaction
+			wrapper.accountQueue.logger.Debug(
+				fmt.Sprintf(
+					"Dropping txn [%s] due to a low nonce [%d < %d]",
+					lowestNonceTx.Hash.String(),
+					lowestNonceTx.Nonce,
+					accountNextNonce,
+				),
+			)
+			wrapper.accountQueue.txs.Pop()
+
+			txnDropCleanup(lowestNonceTx)
+		}
+
+		wrapper.unlock()
+
+		// Prune out all the now possibly low-nonce transactions in the promoted queue
+		t.pendingQueue.lock.Lock()
+
+		// Find the txns that correspond to this account
+		droppedPendingTxs := 0
+		for _, pendingQueueTxn := range t.pendingQueue.index {
+			// Check if the txn in the promoted queue matches the search criteria
+			if pendingQueueTxn.from == txn.From && // The sender of this txn is the account we're looking for
+				pendingQueueTxn.tx.Nonce < accountNextNonce { // The nonce on this promoted txn is invalid
+				// Transaction found, drop it from the pending queue
+				if dropped := t.pendingQueue.dropTx(pendingQueueTxn.tx); dropped {
+					// Update the log data
+					droppedPendingTxs++
+					wrapper.accountQueue.logger.Debug(
+						fmt.Sprintf(
+							"Dropping promoted txn [%s]",
+							pendingQueueTxn.tx.Hash.String(),
+						),
+					)
+
+					txnDropCleanup(pendingQueueTxn.tx)
+				}
+			}
+		}
+
+		t.pendingQueue.lock.Unlock()
```

Comment:
Shall we create a method for this? Are you deleting transactions with lower nonce from pending queue, right?

Reply:
I do agree with @Kourin1996 , it would be cleaner to have a method to remove tx from pending queue.

Reply:
Extracted the logic 👍 

ecc07d8